src/UCTNode.cpp

```diff
@@ -398,6 +398,50 @@ const std::vector<UCTNode::node_ptr_t>& UCTNode::get_children() const {
     return m_children;
 }
 
+int UCTNode::count_nodes() const {
+    auto nodecount = 0;
+    if (m_has_children) {
+        nodecount += m_children.size();
+        for (auto& child : m_children) {
+            nodecount += child->count_nodes();
+        }
+    }
+    return nodecount;
+}
+
+// Use this version if you know the child is directly under the parent.
+UCTNode::node_ptr_t UCTNode::find_and_take_ownership(const int move) {
+    if (m_has_children) {
+        // TODO: I couldn't get the find_if version to work.
+        // auto& child = std::find_if(begin(m_children), end(m_children),
+        //     [&move](auto &child) { return child->get_move == move; });
+        for (auto& child : m_children) {
+            if (child->get_move() == move) {
+                return std::move(child);
+            }
+        }
+    }
+    return nullptr;
+}
+
+// Use this version if the child could be anywhere.
```

Comment:
Not sure I understand in which case this one is useful? I think it's more that if you don't have the move? But get_last_move() will have it...

Reply:
GTP calls this one and I check the hashes to guarantee the entire position is match not just the last move. This covers cases where GTP gets multiple play commands followed by a genmove, without requiring GTP to do extra bookkeeping.

The other version should only be called by UCTSearch::think. In this case we can guarantee the child is a directly under the parent. We also don't have two separate GameStates to do a hash compare anyways.